Blazor RC2 Updates 8.0 (#30576)
dotnet/AspNetCore.Docs: aspnetcore/blazor/forms-and-input-components.md

A project created from the Blazor project template includes the namespace by default in the app's `_Imports.razor` file, which makes the namespace available to the app's Razor components.

:::moniker range=">= aspnetcore-8.0"

> ## Enhanced navigation and form handling

Reviewer: This article is about forms, not navigation, so I think we can remove navigation from this heading.

```suggestion
## Enhanced form handling
```

I also think this section should go much later in the article, probably near the end. Enhanced form handling is totally optional and we don't want to scare off new users with these details right from the start. We can teach them all they need to know about form without any mention of enhanced form handling and then at the end let them know that they can ✨enhance✨ their Blazor forms by adding the appropriate attribute and explain what that does and how it interacts with redirects to non-Blazor pages.

Reviewer: I think we can replace "enhanced navigation from a form" with "enhanced form handling" in this section.

> Examples adopt [enhanced navigation](xref:blazor/fundamentals/routing#enhanced-navigation-and-form-handling) for form POST requests with the `Enhance` parameter for `EditForm` forms or the `data-enhance` attribute for HTML forms (`<form>`):

Reviewer: Yeah, I think we should flip this around and have most of the examples not be enhanced and then just have one section that explains how you can choose to enhance your form if you want to.

```razor
<EditForm Enhance ...>
    ...
</EditForm>
```

```html
<form data-enhance ...>
    ...
</form>
```

> <span aria-hidden="true">❌</span><span class="visually-hidden">Unsupported:</span> You can't set enhanced navigation on a form's ancestor element to enable enhanced navigation for the form.

Reviewer: ```suggestion
<span aria-hidden="true">❌</span><span class="visually-hidden">Unsupported:</span> You can't set enhanced navigation on a form's ancestor element to enable enhanced form handling.
```

```html
<div data-enhance>
    <form ...>
        <!-- NOT enhanced -->
    </form>
</div>
```

> [!WARNING]
> ***Don't try to enhance form navigation for forms that POST to non-Blazor endpoints.***
>

> > This is the primary reason why enabling form POSTs with enhanced navigation requires an explicit gesture on each `EditForm` and `<form>` element. It isn't safe to assume that all forms in general should use enhanced posts because any given form might not target a Blazor endpoint.

Reviewer: I think this warning might sound a bit too scary. I think all we really need to say is that enhanced forms only work with Blazor. Posting an enhanced form to non-Blazor endpoint will result in an error.

> To disable enhanced navigation for a form:

Reviewer: ```suggestion
To disable enhanced form handling:
```

* For an `EditForm`, remove the `Enhance` parameter from the form element (or set it to `false`: `Enhance="false"`).
* For an HTML `<form>`, remove the `data-enhance` attribute from form element (or set it to `false`: `data-enhance="false"`).

> Blazor's enhanced navigation and form handing may undo dynamic changes to the DOM if the updated content isn't part of the server rendering. To preserve the content of an element, use the `data-permanent` attribute.

Reviewer: I don't think we need to duplicate this content here. You've already got it covered in the Blazor routing doc, right? Can we just link to that for these details?

In the following example, the content of the `<div>` element is updated dynamically by a script when the page loads:

```html
<div data-permanent>
    ...
</div>
```

To disable enhanced navigation and form handling globally, see <xref:blazor/fundamentals/startup#enhanced-navigation-and-form-handling>.

Once Blazor has started on the client, you can use the `enhancedload` event to listen for enhanced page updates. This allows for re-applying changes to the DOM that may have been undone by an enhanced page update.

```javascript
Blazor.addEventListener('enhancedload', () => console.log('Enhanced update!'));

> ```

Reviewer: I don't think we need to duplicate this content here. You've already got it covered in the Blazor routing doc, right? Can we just link to that for these details?

:::moniker-end

## Examples in this article

:::moniker range=">= aspnetcore-8.0"

> Components are configured for interactivity with server rendering and enhanced navigation. For a client-side experience in a Blazor Web App, change the render mode in the `@rendermode` directive at the top of the component to either:

Reviewer: I think what we eventually want to do here is build up the user's understanding of form handling in Blazor by building off of concepts they probably already know, like the basics of forms in HTML. So, I think that means the initially sections should start with just a plan `<form>` (not `EditForm`) and then build up from there. First show how you can handle the form submission from the server using `@onsubmit`. Then introduce `EditForm` as a more convenient way to build a server based form. Then introduce enhanced form handling from the server. Then introduce interactive form handling. But we don't have to do this restructuring right away. It can wait until we get the .NET 8 stuff covered.

Author: Very well, but it's probably best to reorganize the coverage after RC2 releases, perhaps late next week. I'll open an issue to make sure the task doesn't get lost when this PR merges.

<!-- UPDATE 8.0 These simplify at RTM -->

* `RenderMode.InteractiveWebAssembly` for interactive client rendering only.
* `RenderMode.InteractiveAuto` for interactive client rendering after interactive server rendering, which operates while the Blazor app bundle downloads in the background and the .NET WebAssembly runtime starts on the client.

If working with a standalone Blazor WebAssembly app, take ***either*** of the following approaches:

* Change the render mode to `RenderMode.InteractiveWebAssembly`:

  ```diff
  - @rendermode RenderMode.InteractiveServer

>   + @rendermode RenderMode.InteractiveWebAssembly

Reviewer: Standalone Blazor WebAssembly apps don't use the render modes. The whole app is running interactively on WebAssembly, just like it always has. All of the interactive forms in this doc should work just fine in a standalone Blazor WebAssembly app as long as the code doesn't make any assumptions about running on the server instead of the client.

<!-- UPDATE 8.0 Cross-link SSR -->

> Standard interactive HTML forms with server rendering are supported without using an <xref:Microsoft.AspNetCore.Components.Forms.EditForm>, including with ENHANCED NAV.

Reviewer: ```suggestion
Standard interactive HTML forms with server rendering are supported without using an <xref:Microsoft.AspNetCore.Components.Forms.EditForm>, including with enhanced form handling.
```

To submit a form based on another element's DOM events, for example `oninput` or `onblur`, use JavaScript to submit the form ([`submit` (MDN documentation)](https://developer.mozilla.org/docs/Web/API/HTMLFormElement/submit)).

> Enhanced navigation is supported with the *explicit addition* of the `data-enhance` HTML attribute.

Reviewer: ```suggestion
Enable enhanced form handling by adding the `data-enhance` HTML attribute to the form.
```

<span aria-hidden="true">❌</span><span class="visually-hidden">Unsupported:</span> You can't set enhanced navigation on a form's ancestor element to enable enhanced navigation for the form.

```html
<div data-enhance>
    <form ...>
        <!-- NOT enhanced -->
    </form>
</div>
```

> [!WARNING]
> ***Don't try to enhance form navigation for forms that POST to non-Blazor endpoints.***
>

> > This is the main reason why enabling form POSTs with enhanced navigation requires an explicit gesture on each `<form>` element. It isn't safe to assume that all forms in general should use enhanced posts because any given form might not target a Blazor endpoint.

Reviewer: Same comment about this warning as the one above. I want to avoid making it sound like enhanced form handling is somehow "unsafe".
